### PR TITLE
Create DNS sub-zone in cluster.dev domain and delegate it

### PR DESCRIPTION
**Pull request by voatsap**
Each cluster should have own dns zone. 

So we could: 
- create multiple subdomains
- use dns certificate validation with cert-manager
- ability to manage zone from Kubernetes external DNS
- detect the service origin by domain
- limit ability to manage records from cluster inside zone

yaml option:
```yaml
cluster
  cloud
    domain:
       own domain -> creates zone CLUSTER_FULLNAME.cluster_cloud_domain
       domain/cluster.dev -> create zone  CLUSTER_FULLNAME.cluster.dev and invoke zone delegation
```

### CODE REVIEW

File: bin/aws_common.sh [hunk 1]
```diff
@@ -86,32 +86,39 @@ function aws::destroy_s3_bucket {
...
-# Outputs:
-#   Writes progress status
 #######################################
 function aws::init_route53 {
     DEBUG "Create a DNS domains/records if required"
+    local default_domain="cluster.dev"
     local cluster_cloud_region=$1
-    local cluster_name=$2
-    local cluster_cloud_domain=$3
+    local CLUSTER_FULLNAME=$2
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:95]: local redefined GLOBAL?

File: bin/aws_common.sh [hunk 2]
```diff
@@ -86,32 +86,39 @@ function aws::destroy_s3_bucket {
...
+        -backend-config="key=$cluster_name/terraform-dns.state" \
+        -backend-config="region=$cluster_cloud_region"
+
+    # Create or update zone
+    if [ "$cluster_cloud_domain" = "$default_domain" ]; then
+            INFO "The cluster domain is unset. DNS sub-zone would be created in $default_domain"
+            zone_delegation=true
+        else
+            INFO "The cluster domain defined. DNS sub-zone would be created in $cluster_cloud_domain"
+            zone_delegation=false
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:110]:
```suggestion
    else
        INFO "The cluster domain defined. DNS sub-zone would be created in $cluster_cloud_domain"
        zone_delegation=false
```

File: bin/aws_common.sh [hunk 3]
```diff
@@ -86,32 +86,39 @@ function aws::destroy_s3_bucket {
...
+    # Execute terraform
+        run_cmd "terraform plan -compact-warnings \
+                -var='region=$cluster_cloud_region' \
+                -var='cluster_fullname=$CLUSTER_FULLNAME' \
+                -var='cluster_domain=$cluster_cloud_domain' \
+                -var='zone_delegation=$zone_delegation' \
+                -input=false \
+                -out=tfplan"
+        run_cmd "terraform apply -auto-approve -compact-warnings -input=false tfplan"
+        INFO "DNS Zone: $CLUSTER_FULLNAME.$cluster_cloud_domain has been created."
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:121]:
```suggestion
    run_cmd "terraform plan -compact-warnings \
            -var='region=$cluster_cloud_region' \
            -var='cluster_fullname=$CLUSTER_FULLNAME' \
            -var='cluster_domain=$cluster_cloud_domain' \
            -var='zone_delegation=$zone_delegation' \
            -input=false \
            -out=tfplan"
    run_cmd "terraform apply -auto-approve -compact-warnings -input=false tfplan"
    INFO "DNS Zone: $CLUSTER_FULLNAME.$cluster_cloud_domain has been created."
```

File: bin/aws_common.sh [hunk 4]
```diff
@@ -123,19 +130,30 @@ function aws::init_route53 {
...
-# Outputs:
-#   Writes progress status
 #######################################
 function aws::destroy_route53 {
-    DEBUG "Destroy a DNS domains/records if required"
+    local default_domain="cluster.dev"
     local cluster_cloud_region=$1
-    local cluster_name=$2
-    local cluster_cloud_domain=$3
+    local CLUSTER_FULLNAME=$2
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:138]: Same

File: bin/aws_common.sh [hunk 5]
```diff
@@ -123,19 +130,30 @@ function aws::init_route53 {
...
+    terraform init -backend-config="bucket=$S3_BACKEND_BUCKET" \
+        -backend-config="key=$cluster_name/terraform-dns.state" \
+        -backend-config="region=$cluster_cloud_region"
+
+    # Execute terraform
+    INFO "Destroying a DNS zone $CLUSTER_FULLNAME.$cluster_cloud_domain"
+        run_cmd "terraform  destroy -auto-approve  \
+                -var='region=$cluster_cloud_region' \
+                -var='cluster_domain=$cluster_cloud_domain' \
+                -var='cluster_fullname=$CLUSTER_FULLNAME'"
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:152]:
```suggestion
    run_cmd "terraform  destroy -auto-approve  \
            -var='region=$cluster_cloud_region' \
            -var='cluster_domain=$cluster_cloud_domain' \
            -var='cluster_fullname=$CLUSTER_FULLNAME'"
```

File: bin/aws_common.sh [hunk 6]
```diff
@@ -123,19 +130,30 @@ function aws::init_route53 {
...
 
-    # TODO: destroy procedure.
+        INFO "DNS Zone: $CLUSTER_FULLNAME.$cluster_cloud_domain has been deleted."
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:154]:
```suggestion
    INFO "DNS Zone: $CLUSTER_FULLNAME.$cluster_cloud_domain has been deleted."
```

File: bin/aws_common.sh [hunk 7]
```diff
@@ -290,7 +308,7 @@ function aws::destroy {
             aws::minikube::destroy_cluster "$cluster_name" "$cluster_cloud_region" "$cluster_cloud_provisioner_instanceType" "$cluster_cloud_domain"
             # TODO: Remove kubeconfig after successful cluster destroy
             aws::destroy_vpc "$cluster_cloud_vpc" "$cluster_name" "$cluster_cloud_region"
-            aws::destroy_route53 "$cluster_cloud_region" "$cluster_name" "$cluster_cloud_domain"
+            aws::destroy_route53 "$cluster_cloud_region" "$CLUSTER_FULLNAME" "$cluster_cloud_domain"
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:311]: Same

File: bin/aws_minikube.sh [hunk 8]
```diff
@@ -86,7 +86,7 @@ function aws::minikube::deploy_cluster {
                 -var='region=$cluster_cloud_region' \
                 -var='cluster_name=$CLUSTER_FULLNAME' \
                 -var='aws_instance_type=$cluster_cloud_provisioner_instanceType' \
-                -var='hosted_zone=$cluster_cloud_domain' \
+                -var='hosted_zone=$CLUSTER_FULLNAME.$cluster_cloud_domain' \
```
Comment (MaxymVlasov) [inline bin/aws_minikube.sh:89]: It accidentally?

File: bin/aws_minikube.sh [hunk 9]
```diff
@@ -132,7 +132,7 @@ function aws::minikube::destroy_cluster {
                 -var='region=$cluster_cloud_region' \
                 -var='cluster_name=$CLUSTER_FULLNAME' \
                 -var='aws_instance_type=$cluster_cloud_provisioner_instanceType' \
-                -var='hosted_zone=$cluster_cloud_domain'"
+                -var='hosted_zone=$CLUSTER_FULLNAME.$cluster_cloud_domain'"
```
Comment (MaxymVlasov) [inline bin/aws_minikube.sh:135]: Same

File: terraform/aws/route53/main.tf [hunk 10]
```diff
@@ -2,19 +2,24 @@ provider "aws" {
   region = var.region
 }
 
-resource "aws_route53_zone" "main" {
-  name = var.cluster_domain
+# If DNS zone is provided by user.
+# Use it to create subzone
+data "aws_route53_zone" "existing" {
+  count = "${var.zone_delegation == "false" ? 1 : 0}"
```
Comment (MaxymVlasov) [inline terraform/aws/route53/main.tf:8]: TF 0.11...

File: terraform/aws/route53/main.tf [hunk 11]
```diff
@@ -23,3 +28,24 @@ resource "aws_route53_record" "sub-ns" {
     "${aws_route53_zone.sub.name_servers.3}",
   ]
```
Comment (MaxymVlasov) [inline terraform/aws/route53/main.tf:29]: Try use
```terraform
records = aws_route53_zone.sub.name_servers
```
or
```terraform
records = [ aws_route53_zone.sub.name_servers.* ]
```

Comment (MaxymVlasov) [reply terraform/aws/route53/main.tf:29]: Same for outputs

File: bin/aws_common.sh (hunk 1 above)
Comment (voatsap) [reply bin/aws_common.sh]: fixed

File: bin/aws_common.sh (hunk 4 above)
Comment (voatsap) [reply bin/aws_common.sh]: fixed

File: bin/aws_common.sh (hunk 7 above)
Comment (voatsap) [reply bin/aws_common.sh:311]: fixed

File: terraform/aws/route53/main.tf (hunk 10 above)
Comment (voatsap) [reply terraform/aws/route53/main.tf]: Fixed

**Comment by voatsap** [comment]
Cluster up test: https://github.com/voatsap/testrepo/runs/625852991?check_suite_focus=true
Cluster down test: https://github.com/voatsap/testrepo/runs/626054284?check_suite_focus=true

File: bin/aws_common.sh [hunk 12]
```diff
@@ -86,32 +86,39 @@ function aws::destroy_s3_bucket {
...
-    if [ -z "$cluster_cloud_domain" ]; then
-        INFO "The cluster domain is unset. It is going to be created default"
+    terraform init -backend-config="bucket=$S3_BACKEND_BUCKET" \
+        -backend-config="key=$cluster_name/terraform-dns.state" \
+        -backend-config="region=$cluster_cloud_region"
+
+    # Create or update zone
+    if [ "$cluster_cloud_domain" = "$default_domain" ]; then
+            INFO "The cluster domain is unset. DNS sub-zone would be created in $default_domain"
+            zone_delegation=true
```
Comment (MaxymVlasov) [inline bin/aws_common.sh:107]:
```suggestion
        INFO "The cluster domain is unset. DNS sub-zone would be created in $default_domain"
        zone_delegation=true
```

File: terraform/aws/route53/main.tf [hunk 13]
```diff
@@ -2,24 +2,45 @@ provider "aws" {
   region = var.region
 }
 
-resource "aws_route53_zone" "main" {
-  name = var.cluster_domain
+# If DNS zone is provided by user.
+# Use it to create subzone
+data "aws_route53_zone" "existing" {
+  count = tobool(var.zone_delegation) ? 0 : 1
+  name         = "${var.cluster_domain}."
```
Comment (MaxymVlasov) [inline terraform/aws/route53/main.tf:9]: Yay! `terraform fmt` bored without you

File: bin/aws_common.sh (hunk 7 above)
Comment (MaxymVlasov) [inline bin/aws_common.sh:311]: Please, write funtion doc, like

```bash
#######################################
# Writes commands for user for get access to cluster
# Globals:
#   CLUSTER_FULLNAME
# Arguments:
#   cluster_cloud_domain
# Outputs:
#   Writes commands to get cluster's kubeconfig and ssh key
#######################################
```

File: terraform/aws/route53/main.tf (hunk 13 above)
Comment (voatsap) [reply terraform/aws/route53/main.tf]: fixed

File: bin/aws_common.sh (hunk 7 above)
Comment (voatsap) [reply bin/aws_common.sh:311]: It was done before

#######################################
# Destroy a DNS domains/records if required
# Globals:
#   S3_BACKEND_BUCKET
#   CLUSTER_FULLNAME
# Arguments:
#   cluster_cloud_region
#   cluster_cloud_domain
#######################################

**Pull request merged**